flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NettyShuffleEnvironment.java

```diff
@@ -184,17 +184,18 @@ public ShuffleIOOwnerContext createShuffleIOOwnerContext(
 	}
 
 	@Override
-	public Collection<ResultPartition> createResultPartitionWriters(
```

Comment:
I think we do not really need this commit change for guaranteeing the consistent info during recovery.  There should be no order concerns inside the collection, and we only need to guarantee that the generated `ResultPartition` and `SingleInputGate` inside the collection has always consistent infos as before. And this guarantee has already been realized by making `ResultPartitionDeploymentDescriptor` as list.

Reply:
Hm seems like my response got lost.
In general, `Collection` gives no ordering guarantees. If the using code breaks if we return a `HashSet` instead of an `ArrayList`, then it's the wrong interface. We have been sloppy with this distinction in the past and I wanted to make it very explicit for our deep change into how channels are working.

Reply:
Yes, your above comments are exactly right. 

There were no ordering requirements when this interface was introduced firstly.  Considering the recovery issue of unaligned checkpoint, the ordering issue was already guaranteed while constructing the collections of `ResultPartition` via argument `List<ResultPartitionDeploymentDescriptor>`, done by a separate PR before.  

Now even though we return a `HashSet` as a result, it would not impact the unaligned checkpoint. Because as long as we create the `ResultPartition` in determined index, it does not matter to access them in any order from the collection.

I am also fine if you further want to guarantee the order from return type, for the potential future usages, but I guess that is not the blocker for current unaligned checkpoint.

Reply:
But current code implicitly assumes the partition indexes when calling `Collection#toArray` in `Task` class:
```
// produced intermediate result partitions
final ResultPartitionWriter[] resultPartitionWriters = shuffleEnvironment.createResultPartitionWriters(
	taskShuffleContext,
	resultPartitionDeploymentDescriptors).toArray(new ResultPartitionWriter[] {});
```
so I think it indeed should be a `List`, not a `Collection`.